fix: Detect host and protocol for alternate links correctly when running behind a proxy

HHongSeungWoo opened the pull request:
We are currently utilizing a reverse proxy in front of Next.js.
`request.url` is returning an internal address instead of an external one.
Consequently, the rewrite functionality is not behaving as expected, and the `alternatelink` header is inadvertently exposing the internal address.

HHongSeungWoo commented:
I'm concerned about trusting the host of the header, but if there are multiple external addresses on the same server, I can't think of any other solution :(

amannn commented on packages/next-intl/src/middleware/middleware.tsx:
```diff
@@ -32,6 +33,9 @@ export default function createMiddleware(config: MiddlewareConfig) {
   return function middleware(request: NextRequest) {
+    // if Next.js behind a reverse proxy, it cannot get the correct host. ex) localhost, 172.x.x.x
+    // We need to set the host here, because Next.js doesn't do it for us
+    request.nextUrl.host = getHost(request.headers) ?? request.nextUrl.host;
```
If possible, I'd like to avoid overwriting the incoming request. If the user [composes the middleware](https://next-intl-docs.vercel.app/docs/routing/middleware#composing-other-middlewares), this could lead to surprises.

HHongSeungWoo replied:
This seems to be a new issue arising in Next.js version ^13.4.13. Since it might not be a problem that needs to be addressed within this library, I will remove it for now.

HHongSeungWoo replied:
In previous versions, the host of request.url was pointing to localhost, allowing for correct rewriting or redirection. However, it seems that the behavior has changed in version ^13.4.13.

amannn replied:
Interesting. Do you have a link to the corresponding change or a bug report on the Next.js side by chance?

HHongSeungWoo replied:
It seems like this might be related to https://github.com/vercel/next.js/issues/54133.

amannn commented on packages/next-intl/src/middleware/middleware.tsx:
```diff
@@ -68,15 +72,18 @@ export default function createMiddleware(config: MiddlewareConfig) {
+      return NextResponse.rewrite(
+        new URL(url, request.nextUrl.href),
+        getResponseInit()
+      );
```
Is it necessary to rewrite to the external host here (same for the redirect)?

In your PR comment you mentioned:
> Consequently, the rewrite functionality is not behaving as expected

Can you explain what is not behaving as expected? My impression would be that your Next.js app already receives requests on an internal host, why can't it handle rewrites to the internal address too?

That being said, your point is true that the alternate links should point to the external host. I'm just wondering if we could limit the fix to choosing the correct host in `getAlternateLinksHeaderValue`, not having to modify anything outside of that file.

What do you think?

HHongSeungWoo replied:
I haven't read about what alternatelink does exactly, but I'm curious about why the domain needs to be included. Can't the desired functionality be achieved with just the pathname without including the domain?

amannn replied:
As far as I know, the domain needs to be included: https://developers.google.com/search/docs/specialty/international/localized-versions#http

amannn reviewed (commented):
Thank you so much for proposing this fix @HHongSeungWoo!

It's definitely a good idea to use the external host for the alternate links, please see my inline comment for details.

Can you please also include tests that would have failed previously and will now pass after the fix?

Thank you so much! 🙌

HHongSeungWoo commented:
![스크린샷 2023-08-21 18-23-30](https://github.com/amannn/next-intl/assets/34399997/df9bed2f-0135-41f1-b5c6-14e1df385db5)
![스크린샷 2023-08-21 18-22-31](https://github.com/amannn/next-intl/assets/34399997/b5a6d95f-48f9-4b98-9254-79aca96160ed)

before / after

dzenand commented:
Is this somehow related to this issue? 
https://github.com/vercel/next.js/pull/54277

amannn commented:
This looks great! The only thing missing would be a regression test that helps to ensure that we keep this functionality in the future.

Thanks also for linking to the Next.js issue, I'll keep an eye on that, but seems like it should be addressed on the Next.js side.

HHongSeungWoo commented:
> This looks great! The only thing missing would be a regression test that helps to ensure that we keep this functionality in the future.
> 
> Thanks also for linking to the Next.js issue, I'll keep an eye on that, but seems like it should be addressed on the Next.js side.


I'm unsure about the necessary test cases. The request.url within the middleware is a string generated by Next.js. 
If a hostname is configured, it will be used otherwise, localhost will be used. Upon reviewing the code, it appears that the option trustHostHeader, which is not currently utilized, generates the URL with the request's host when set to true. However, since this isn't documented, it seems to be an unused option. These changes are independent of external configurations or code modifications, so I can't think of any regression cases unless there are changes to the HTTP protocol itself.

amannn commented:
@HHongSeungWoo I've added a commit where the now fixed behavior is handled within a single test case. I also noticed that we should consider the protocol as well and fixed that accordingly.

Thanks for noticing that we can use the `NextRequest` constructor instead of constructing a mock object, I've adapted the tests to use that consistently to avoid unnecessary redirection.

Would you like to have a last look? From my opinion this is good to go! 👍

HHongSeungWoo commented:
> @HHongSeungWoo I've added a commit where the now fixed behavior is handled within a single test case. I also noticed that we should consider the protocol as well and fixed that accordingly.
> 
> Thanks for noticing that we can use the `NextRequest` constructor instead of constructing a mock object, I've adapted the tests to use that consistently to avoid unnecessary redirection.
> 
> Would you like to have a last look? From my opinion this is good to go! 👍


Thank you for making the necessary corrections. It looks better to me as well. :)

CarbonC commented:
Hello @amannn ! This looks great, but please would it be possible to push this fix to the beta branch as well? Currently im stuck with the bug on the beta-10 :-/
Thanks for the great work

amannn commented:
@CarbonC Yep, definitely! I've got another release on the main branch coming, then I'll rebase the RSC branch and publish a new beta!

amannn commented:
@CarbonC `next-intl@3.0.0-beta.11` is out now with all recent improvements from the main branch!
